Use dynamicDowncast in MathMLSelectElement

AlexsanderDamaceno opened the pull request:
#### e73a61ac0456037efd88e15bbbfa2d2b2e80b826
<pre>
Use dynamicDowncast in MathMLSelectElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=281559">https://bugs.webkit.org/show_bug.cgi?id=281559</a>

Reviewed by Chris Dumez and Sihui Liu.

Use dynamicDowncast in MathMLSelectElement

* Source/WebCore/mathml/MathMLSelectElement.cpp:
(WebCore::MathMLSelectElement::getSelectedSemanticsChild):

Canonical link: <a href="https://commits.webkit.org/285307@main">https://commits.webkit.org/285307@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6dc3fa42c7785a583c4156c4865812b408392618

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72102 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51523 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24887 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76267 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23311 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74217 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59333 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23131 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56873 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15382 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75169 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46714 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62111 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37313 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43375 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19585 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21661 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65280 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19945 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77947 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16343 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19111 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/65345 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16390 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62134 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64603 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15950 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12806 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6457 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47320 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2105 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48390 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49678 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48133 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebCore/mathml/MathMLSelectElement.cpp:
```diff
@@ -164,7 +164,7 @@ Element* MathMLSelectElement::getSelectedSemanticsChild()
         return nullptr;
 
-    if (!is<MathMLElement>(*child) || !downcast<MathMLElement>(*child).isPresentationMathML()) {
+    if (!dynamicDowncast<MathMLElement>(*child)->isPresentationMathML()) {
```
`dynamicDowncast` might return null when child is not `MathMLElement`, and this will be null dereference

AlexsanderDamaceno replied:
@szewai but In this case I think  

 !is<MathMLElement>(*child) || !downcast<MathMLElement>(*child).isPresentationMathML() is checking for the null condition for enter the  if block, no?

cdumez replied:
This can be better written as:
```
if (auto* childElement = dynamicDowncast<MathMLElement>(*child); childElement && childElement->isPresentationMathML()) {
```

This is what Sihui meant and she's right. It avoid doing the type check twice.

AlexsanderDamaceno replied:
@cdumez Ah, ok, thank you.

szewai reviewed (changes requested):
(Clicked Approve by mistake...)

szewai commented on Source/WebCore/mathml/MathMLSelectElement.cpp:
```diff
@@ -164,7 +165,7 @@ Element* MathMLSelectElement::getSelectedSemanticsChild()
         return nullptr;
 
-    if (!is<MathMLElement>(*child) || !downcast<MathMLElement>(*child).isPresentationMathML()) {
+    if (auto* childElement = dynamicDowncast<MathMLElement>(*child); childElement && childElement->isPresentationMathML()) {
```
Still not quite right; it should be `or`, not `and` to keep the old behavior like:
`if (auto* childElement = dynamicDowncast<MathMLElement>(*child); !childElement || childElement->isPresentationMathML())`

cdumez replied:
I may have typed it wrong in my earlier review comment, if so, I apologize.

But @szewai still typed it wrong again I think :p, both conditions should use `!`:
```
if (auto* childElement = dynamicDowncast<MathMLElement>(*child); !childElement || !childElement->isPresentationMathML()) {
```

szewai replied:
ah nice catch; not sure the new change is better then since it's so easy to mistype : P!

AlexsanderDamaceno replied:
Right correctly, thanks @szewai @cdumez .

szewai commented on Source/WebCore/mathml/MathMLSelectElement.cpp:
```diff
@@ -1,3 +1,4 @@
+
```
Unnecessary blank line

AlexsanderDamaceno committed:
Use dynamicDowncast in MathMLSelectElement
https://bugs.webkit.org/show_bug.cgi?id=281559

Reviewed by Chris Dumez and Sihui Liu.

Use dynamicDowncast in MathMLSelectElement

* Source/WebCore/mathml/MathMLSelectElement.cpp:
(WebCore::MathMLSelectElement::getSelectedSemanticsChild):

Canonical link: https://commits.webkit.org/285307@main
